packages/protocol/contracts/L1/provers/GuardianProver.sol

```diff
@@ -0,0 +1,117 @@
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { EssentialContract } from "../../common/EssentialContract.sol";
+import { Proxied } from "../../common/Proxied.sol";
+
+import { LibTiers } from "../tiers/ITierProvider.sol";
+import { TaikoData } from "../TaikoData.sol";
+
+/// @title GuardianProver
+contract GuardianProver is EssentialContract {
+    uint256 public constant MAX_GUARDIANS = 5;
+    uint256 public constant REQUIRED_GUARDIANS = 3;
+    uint256 private constant DONE = type(uint256).max;
+
+    mapping(uint256 id => address guardian) public guardians;
+    mapping(bytes32 => uint256 approvalBits) public blocks;
+
+    uint256[48] private __gap;
+
+    error INVALID_GUARDIAN_SET();
+    error INVALID_GUARDIAN();
+    error INVALID_PROOF();
+    error PROVE_FAILED(bytes);
+
+    /// @notice Initializes the contract with the provided address manager.
+    /// @param _addressManager The address of the address manager contract.
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    /// @notice Adds guardians
+    /// @param multisigParticipants The address array of the guardians.
+    function addGuardians(address[] memory multisigParticipants)
+        external
+        onlyOwner
+    {
+        // Always send 5 addresses, even if you just want to modify 1 address,
+        // so that (Bob, 0, 0, 0, 0) will set Bob only with index 1.
+        if (multisigParticipants.length != MAX_GUARDIANS) {
+            revert INVALID_GUARDIAN_SET();
+        }
+
+        for (uint256 i = 1; i <= MAX_GUARDIANS; i++) {
+            guardians[i] = multisigParticipants[i - 1];
+        }
+    }
+
+    /// @dev Called by each of the 5 guardians
+    function proveBlock(
+        uint64 blockId,
+        TaikoData.BlockEvidence memory evidence,
+        bool unprovable
+    )
+        external
+    {
+        if (
+            evidence.proof.length != 0
+                && evidence.tier != LibTiers.TIER_GUARDIAN
+        ) revert INVALID_PROOF();
+        bytes32 hash = keccak256(abi.encode(blockId, evidence, unprovable));
+
+        uint256 approvalBits = blocks[hash];
+        require(approvalBits != DONE);
+
+        approvalBits |= uint8(1 << getGuardianId(msg.sender));
+
+        if (isApproved(approvalBits)) {
+            if (unprovable) {
+                evidence.proof =
+                    bytes.concat(bytes32(keccak256("RETURN_LIVENESS_BOND")));
+            }
+
+            bytes memory data = abi.encodeWithSignature(
+                "proveBlock(uint64,bytes)", blockId, abi.encode(evidence)
+            );
+
+            (bool success, bytes memory result) =
+                resolve("taiko", false).call(data);
+            if (!success) {
+                revert PROVE_FAILED(result);
+            }
+            blocks[hash] = DONE;
```

Comment:
maybe we can remove `blocks[hash] = DONE;` and use `delete blocks[hash]`.

Reply:
I think `delete` is not 100% good, because it does not signal if a transition is DONE or not. (Or done but deleted). I'd keep it as is.

Reply:
What about we emit an event instead of having `blocks[hash] = DONE;`